#' Retrieve chromosomal length information from file
#'
#' Retrieves from a genome file (alternatively, a FASTA index file generated by samtools) a list of chromosomes and their corresponding lengths. If no input is specified, built-in hg19 information is returned as a data.frame.
#' @param genome filename of the genome file (defaults to NULL); required. If unspecified, hg19 information is returned.
#' @keywords genome chromosome samtools
#' @export

fetch.genome <- function(genome = NULL) {
	if (is.null(genome)) {
		# define with hg19
		hg19 <- data.frame(
			chrom = qw("chr1 chr2 chr3 chr4 chr5 chr6 chr7 chr8 chr9 chr10 chr11 chr12 
			chr13 chr14 chr15 chr16 chr17 chr18 chr19 chr20 chr21 chr22 chrM chrX chrY"),
			ln = qw("249250621 243199373 198022430 191154276 180915260 171115067 159138663 146364022
			141213431 135534747 135006516 133851895 115169878 107349540 102531392 90354753
			81195210 78077248 59128983 63025520 48129895 51304566 16571 155270560 59373566",
			as.is = FALSE), stringsAsFactors = FALSE)
		return(hg19)
	}
	if (file.exists(genome)) {
		gtab <- read.table(file = genome, header = FALSE, as.is = TRUE, sep = "\t")
		return(gtab[,1:2])
	}
	cat("File not found:", genome, "\n")
	return(NA)
}
